api/v3/Payment.php

```diff
@@ -226,14 +235,18 @@ function _civicrm_api3_payment_delete_spec(&$params) {
  *   Array of parameters.
  */
 function _civicrm_api3_payment_cancel_spec(&$params) {
-  $params = array(
-    'id' => array(
+  $params = [
+    'id' => [
       'api.required' => 1 ,
       'title' => 'Payment ID',
       'type' => CRM_Utils_Type::T_INT,
-      'api.aliases' => array('payment_id'),
-    ),
-  );
+      'api.aliases' => ['payment_id'],
+    ],
+    'trxn_date' => [
+      'title' => 'Cancel Date',
+      'type' => CRM_Utils_Type::T_DATE + CRM_Utils_Type::T_TIME,
```

Comment:
Also, trxn_id is important for canceling a payment.

Reply:
@monishdeb logically I agree with you - the current api is not really doing 'Payment.cancel' it's doing 'cancel one payment associated with the contribution' - which really makes me question whether the api is sensible & worth having. That aside, I think this PR doesn't CHANGE the situation in that I think trxn_id is not used currently?

Reply:
ohhh I think the only place Payment.cancel is currently called from is

```
function civicrm_api3_payment_create(&$params) {
  // Check if it is an update
  if (CRM_Utils_Array::value('id', $params)) {
    $amount = $params['total_amount'];
    civicrm_api3('Payment', 'cancel', $params);
    $params['total_amount'] = $amount;
  }
```

That is hurting my brain - I guess the idea is you can't alter a payment only reverse & recreate it?

Reply:
Yes, my earlier point of not having that name as 'cancel' as we are not actually cancelling a existing payment but rather refunding the original payment. Reason why I was suggesting to use refund at it actually adds a new reverse transaction and doesn't affect the original payment. It keeps the clarity in bookkeeping report.

Reply:
I agree PR doesn't change the workflow, I was suggesting to add more info. to this API so a dev. will be aware of what sort of parameter needed to refund a payment. Also ```Cancel Date``` is not appropriate it has to be termed as 'Refund Date' as we are actually refunding the payment. Like you earlier said Cancelling a contribution is something need to dealt at Form level ?

Reply:
@monishdeb so leaving aside the payment processor part & just talking about the payments - we would wind up with

Payment.create
Payment.refund - which would take Payment Id & create a refund payment
Payment.cancel - the value of which is unclear

Reply:
Payement.refund - what about backoffice refund for which we don't use any PP?

Reply:
I am saying not on payment processor aspect but on general refund aspect as a refund can be done without using payment processor like currently Civi is supporting via 'Record Refund' backoffice form.

Reply:
Right - so I'm NOT suggesting that Payment.Refund would call the payment processor refund function.

I'm suggesting we ADD a PaymentProcessor.refund action that ONLY calls the paymentProcessor refund function (& doesn't do any payment stuff)

Reply:
We *might* provide an additional api that calls one & then the other but ideally that would hold two lines of code - ie. the 2 api calls.

If there is no payment_processor_id we'd just load CRM_Core_Payment_Manual as our processor class & call that - so the fn doesn't need to 'care' about that

Reply:
But - I think we get the 2 api calls right & then look at integrating into a form & decide if we DO need a wrapper api to hold them both or if the api should just call them.

Reply:
Hmm that means as per my earlier patch (not a concrete one, just for experimenting purposes with Moneris refund ) The Payment.refund should be PaymentProcessor.refund and eventually to complete a refund payment:
```php
PaymentProcessor.refund($params) {
 $result = $processor->dorefund($params['payment_processor_id']);
 civicrm_api3('Payment', 'cancel', $result);
}
```
Sorry I should not dive more into PP refund aspect. I was just curious how the Payment.cancel API would be involved in PP refund support and this patch is essential in deciding that.

Reply:
@monishdeb So I'm actually suggesting it's even simpler - ie it doesn't call 
```
 civicrm_api3('Payment', 'cancel', $result);
```

e.g here is the PaymentProcessor.pay api from Omnipay
```
/**
 * Action payment.
 *
 * @param array $params
 *
 * @return array
 *   API result array.
 * @throws CiviCRM_API3_Exception
 */
function civicrm_api3_payment_processor_pay($params) {
  $processor = Civi\Payment\System::singleton()->getById($params['payment_processor_id']);
  $processor->setPaymentProcessor(civicrm_api3('PaymentProcessor', 'getsingle', array('id' => $params['payment_processor_id'])));
  $result = $processor->doPayment($params);
  if (is_a($result, 'CRM_Core_Error')) {
    throw API_Exception('Payment failed');
  }
  return civicrm_api3_create_success(array($result), $params);
}
```

The calling code would call 

```
civicrm_api3('PaymentProcessor', 'refund', $params);
 civicrm_api3('Payment', 'create', $result);

```

Whether we want an api that does those 2 lines depends a bit on how it looks as we get further down the track

Reply:
Oops missed the fact that Payment.create does call the Payment.cancel after ensuring its refund payment. I agree with you :)

Reply:
Yeah that's slightly terrifying - part of me is wondering if this PR will create a code loop that will crash the server.... But I guess it didn't before.... 

I feel like this is a BAD function but my goal in this PR is not to address that but just to get to the point where recordAdditionalPayment is only called from one place in the code....

Reply:
Yes maybe I am seeing a bigger picture here to use Paymentprocessor to do partial/pending/refund payment all of which utilizes Payment.create API But that dream is after we simplify and strengthen the Payment.create BAO fn to handle additional payment and of course to get recordAdditionalPayment called from one place

Reply:
Right & maybe once we've one that I'll feel more comfortable about a more complex api - all I can see at the moment is this spaghetti of code trying to do 25 things at once & hence I'm resistant to more multipurpose functions while this is in front of me